### PR TITLE
WIP: A crud generator inspired by Ruby On Rails scaffolding

### PR DESCRIPTION
**Pull request by netodevel**
### This is a demonstration of what I'm developing
_**Attention:
turn down the volume before running the demo.**_

https://user-images.githubusercontent.com/10883521/124366147-e04c6080-dc23-11eb-8f48-1350864c04fd.mp4

**_A crud generator inspired by Ruby On Rails scaffolding._**

### Goal
Automate/generate repetitive daily tasks (or at least the beginning of them) as a simple CRUD.

### What I planned for the first version

Crud generator in Java with standard structure (model, service, controller) and tests.
Maven and Gradle support.

commands:

```
mvn quarkus:model -Dparams="User name:String age:Int createAt:Date"
mvn quarkus:repository -Dparams="User"
mvn quarkus:service -Dparams="User"
mvn quarkus:repository -Dparams="User"
mvn quarkus:scaffold -Dparams="User name:String age:Int createAt:Date"
```
### Future

In the future this may go beyond generating crud and also generating initial settings.

Input:
`mvn quarkus:config -Dparam="amazon-sqs"`

Output:
> add dependencies
> add properties
> add example of consumer and producer

In the future it will be possible to choose which target language, such as: kotlin.
and control this by some properties files that the generator will consult before executing.
I also see generation of migrations(liquibase,flyway) in the future.

### This feature is still on WIP.
I post here to hear opinions about and help me complete this feature as well.
Thanks.

### CODE REVIEW

**Comment by geoand** [comment]
cc @ebullient @maxandersen

**Comment by maxandersen** [comment]
Looks good and interesting; I do though think making this something as part of maven/gradle plugin is not the way forward. Ultimately should be a plugin to quarkus cli. That has the advantage that this wont be dependent on core repository but can be done separately (like quarkus jhipster support is not part of quarkus core).

We don't have the plugin mechanism yet but will soon; follow #18408 to know more.

File: independent-projects/tools/devtools-common/pom.xml [hunk 1]
```diff
@@ -42,6 +42,12 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-devtools-codestarts</artifactId>
         </dependency>
+        <!-- Generator Mustache -->
+        <dependency>
+            <groupId>com.github.spullara.mustache.java</groupId>
```
Comment (maxandersen) [inline independent-projects/tools/devtools-common/pom.xml:47]: a specific reason you don't use qute ?

Comment (NetoDevel) [reply independent-projects/tools/devtools-common/pom.xml:47]: just because I am more familiar with mustache, I have never used qute for this purpose.

**Comment by NetoDevel** [comment]
> Looks good and interesting; I do though think making this something as part of maven/gradle plugin is not the way forward. Ultimately should be a plugin to quarkus cli. That has the advantage that this wont be dependent on core repository but can be done separately (like quarkus jhipster support is not part of quarkus core).
> 
> We don't have the plugin mechanism yet but will soon; follow #18408 to know more.

really, coupling to the core is not a very good thing.
I just discovered the cli. :smile: 
https://quarkus.io/guides/cli-tooling

the idea then is this scaffold to be a separate project and use it as a "plugin" to be installed on the cli, right?


I just have a counterpoint for us to think about.
The java community (that I know of) has no familiarity or desire to install cli for their development workflow.
that's why I adopted engaging in the practicality of using maven/gradle;

I've been through this in the spring community.
I really count on my fingers how many people have already installed the spring cli and used the scaffold plugin.
https://github.com/NetoDevel/cli-spring-boot-scaffold

**Comment by maxandersen** [comment]
> the idea then is this scaffold to be a separate project and use it as a "plugin" to be installed on the cli, right?

basically yes.

@NetoDevel now I don't disagree with you that some devs don't like to have cli's when they just have mvn/gradle plugins...the cool thing is that with the way the cli is made nothing really prevent us from making a quarkus:cli maven and gradle command available. That would be a bit meta and I'm not a superfan of it and there could be a speed penalty but it could be considered.

One thing is though a concern I always have with these codegenerators it is that they go stale if not maintained continuously and thats my main reason to want to decouple that from the core quarkus plugins to let them move as fast or slow as they want/are able to.

**Comment by NetoDevel** [comment]
@maxandersen, 
Okay, given this scenario, I think we can move all the code that makes the generator to the `Quarkiverse Hub`?
When the plugin feature in the cli is ready, maybe this project is a plugin that is already installed by default?
We can go in line to experiment only in CLI, and if adoptions are low maybe try it on maven/gradle?
Did you agree?

_**I really see in the future the quarkus site emphasizing the use of cli with generators to make development fun.
And maybe change this community's idea about cli.**_

**Comment by NetoDevel** [comment]
@ebullient @maxandersen @ia3andy,
Guys, any directions how should I proceed?

**Comment by maxandersen** [comment]
@NetoDevel sorry for the delay - the way forward is you create a repo with a command called quarkus-<whatever you want the plugin root command to be> and make it generate the scaffolding code.

then when we (hopefully soon) have quarkus plugin support it becomes easy to integrate it.

At this time we don't plan on having a "in-core" scaffolding tool - projects like the one you made here, jhipster, nubesgen and others all are available.  If someone standsout as being really fitting we might consider make it the default or extra easy to install, but lets say you make this available in netodvel/jbang-catalog the only thing you would need to do is:

```
jbang app install quarkus-scaffold@netodevel 
quarkus scaffold
```

This way you and contributors can grow and build this on top of Quarkus rather than being tied to specific version of quarkus.

wdyt?

**Comment by ia3andy** [comment]
As @maxandersen said, I would suggest to give a try to Qute for the templates, it's really nice and easy to use (and it is native compatible).

**Comment by NetoDevel** [comment]
@maxandersen, @ia3andy,

I agree, so i will close this pr.
Thanks.

**Pull request closed without merging**